src/AwkwardArray.jl

```diff
@@ -1,5 +1,109 @@
 module AwkwardArray
 
-# Write your package code here.
+abstract type Content end
+
+function Base.iterate(x::Content)
+    start = firstindex(x)
+    stop = lastindex(x)
+    if stop >= start
+        x[start], start + 1
+    else
+        nothing
+    end
+end
+
+function Base.iterate(x::Content, state)
+    stop = lastindex(x)
+    if stop >= state
+        x[state], state + 1
+    else
+        nothing
+    end
+end
+
+### PrimitiveArray #######################################################
+
+struct PrimitiveArray{T} <: Content
+    data::AbstractArray{T,1}
+end
+
+function is_valid(x::PrimitiveArray)
+    true
+end
+
+function Base.length(x::PrimitiveArray)
+    Base.length(x.data)
+end
+
+function Base.firstindex(x::PrimitiveArray)
+    Base.firstindex(x.data)
+end
+
+function Base.lastindex(x::PrimitiveArray)
+    Base.lastindex(x.data)
+end
+
+function Base.getindex(x::PrimitiveArray, i::Int)
+    x.data[i]
+end
+
+function Base.getindex(x::PrimitiveArray, r::UnitRange{Int})
+    PrimitiveArray(x.data[(r.start):(r.stop)])
+end
+
+function Base.:(==)(x::PrimitiveArray, y::PrimitiveArray)
+    x.data == y.data
+end
+
+### ListOffsetArray ######################################################
+
+struct ListOffsetArray{T<:Union{Int32,UInt32,Int64}} <: Content
+    offsets::AbstractArray{T,1}
+    content::Content
+end
+
+function is_valid(x::ListOffsetArray)
+    for i = Base.firstindex(x):Base.lastindex(x)
```

Comment:
```suggestion
    for i = eachindex(x)
```

if `size()` is defined, then `eachindex()` will just work (via `Base.axes()`, which fallsback to `size()`)

Reply:
To use this, I had to define

```julia
function Base.size(x::Content)
    (length(x),)
end
```

But that's a good thing to have because it's required by the [AbstractArray protocol](https://docs.julialang.org/en/v1/manual/interfaces/#man-interface-array).